tower-request-modifier/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+extern crate futures;
+extern crate http;
+extern crate tower_service;
+
+use futures::Poll;
+use http::header::{HeaderName, HeaderValue};
+use http::{Request, HttpTryFrom};
+use http::uri::{self, Authority, Scheme, Uri};
+use tower_service::Service;
+
+#[derive(Debug, Clone)]
+pub enum Modification {
+    SetOrigin(Scheme, Authority),
+    AddHeader(HeaderName, HeaderValue),
+}
+
+/// Wraps an HTTP service, injecting authority and scheme on every request.
+#[derive(Debug, Clone)]
+pub struct RequestModifier<T> {
+    inner: T,
+    modifications: Vec<Modification>,
+}
+
+/// Configure an `RequestModifier` instance
+#[derive(Debug, Default)]
+pub struct Builder {
+    modifications: Vec<Result<Modification, BuilderError>>,
+}
+
+/// Errors that can happen when building an `RequestModifier`.
+#[derive(Debug)]
+pub struct BuilderError {
+    _p: (),
+}
+
+// ===== impl RequestModifier ======
+
+impl<T> RequestModifier<T> {
+    /// Create a new `RequestModifier`
+    pub fn new(inner: T, modifications: Vec<Modification>) -> Self {
+        RequestModifier {
+            inner: inner,
+            modifications: modifications,
+        }
+    }
+
+    /// Return a reference to the modifications that are performed on all requests.
+    pub fn modifications(&self) -> &Vec<Modification> {
```

Comment:
I would probably not include this function in the public API. This ties the struct to the `Vec<Modification>` internal representation. I also don't think we need this.